cwrapper/CArachneWrapperCTest.c

```diff
@@ -0,0 +1,149 @@
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "CArachneWrapper.h"
+
+#include <errno.h>
+#include <assert.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <unistd.h>
+
+enum TestReturn { TEST_SKIP, TEST_PASS, TEST_FAIL };
+typedef enum TestReturn TestReturn;
+
+/**
+ * Functions below are used to test thread creation
+ */
+static void*
+funcCreateTest(void* arg) {
+    usleep(1000);
+    *(unsigned*)arg = 0xDEADBEEF;
+    return NULL;
+}
+
+static TestReturn
+createThreadTest(void) {
+    CArachneThreadId threadId;
+    unsigned arg = 0;
+    TestReturn retval = TEST_PASS;
+
+    cArachneInit(NULL, NULL);
+    int ret = cArachneCreateThread(&threadId, funcCreateTest, (void*)&arg);
+    assert(ret ==  0);
+
+    cArachneJoin(&threadId);
+    if (arg != 0xDEADBEEF) {
+        retval = TEST_FAIL;
+    }
+
+    return retval;
+}
+
+/**
+ * Structures are used to collect all test cases
+ */
+typedef TestReturn (*TEST_FUNC)(void);
+struct testcase {
+    const char *description;
+    TEST_FUNC function;
+};
+
+struct testcase testcases[] = {
+    { "createThread", createThreadTest },
+    { NULL, NULL }
+};
+
+/**
+ * Main entrance for all tests
+ */
+int mainTestLoop(int argc, char **argv) {
+    int exitcode = 0;
+    int id = 0, numCases = 0;
+
+    for (numCases = 0; testcases[numCases].description; numCases++) {
+        /* Counting the number of cases*/
+    }
+
+    printf("Testing cases 1..%d\n", numCases);
+
+    for (id = 0; id < numCases; id++) {
+        fflush(stdout);
+
+        TestReturn ret = testcases[id].function();
+        if (ret == TEST_SKIP) {
+            fprintf(stdout, "ok # SKIP %d - %s\n", id + 1, 
+                   testcases[id].description);
+        } else if (ret == TEST_PASS) {
+            fprintf(stdout, "ok %d - %s\n", id + 1, testcases[id].description);
+        } else {
+            fprintf(stdout, "error %d - %s\n", id + 1, 
+                    testcases[id].description);
+            exitcode = 1;
+            break;
+        }
+        fflush(stdout);
+    }
+
+    return exitcode;
+}
+
+/*
+ * Wrapper for main function arguments
+ */
+struct ArgList {
+    int argc;
+    char** argv;
+};
+typedef struct ArgList ArgList;
+
+/*
+ * Wrapper for real main function
+ */
+static void* mainWrapper(void* arg) {
+    ArgList* mainArgs = (ArgList*)arg;
+    int ret = mainTestLoop(mainArgs->argc, mainArgs->argv);
+    if (ret != 0) {
+        fprintf(stderr, "Non-zero return code: %d\n", ret);
+    }
+    cArachneShutDown();
+    return NULL;
+}
+
+/*
+ * Create main thread
+ */
+int main(int argc, char** argv) {
+    CArachneThreadId threadId;
+    int retval = 0;
+    ArgList mainArgs;
+
+    mainArgs.argc = argc;
+    mainArgs.argv = argv;
+
+    system("../CoreArbiter/bin/coreArbiterServer > /dev/null 2>&1 &");
+    sleep(1); // wait for coreArbiterServer start
```

Comment:
Let's not do this kind of blind wait unless it's a lot of hassle to check whether the server has started.

One option is [popen](http://www.cs.uleth.ca/~holzmann/C/system/shell_commands.html).

Reply:
Now there's no "started message" provided by CoreArbiter server. So it would be impossible to check server status from popen output.

I was thinking to add a started message (printf("CoreArbiter server started!")) in [startArbitration](https://github.com/PlatformLab/CoreArbiter/blob/ebf0979709d5fd4b1511941fd48b2841b2f197fd/src/CoreArbiterServer.cc#L344) function. The startArbitration seems to be the entry point for CoreArbiter.

After that, I can use `popen` to capture the started message.